### ISSUE TITLE
[JIT] Support for CUDA streams in TorchScript

### ISSUE THREAD

**Issue by suo** [pytorch/pytorch#41355]
## 🚀 Feature
Eager mode exposes [a CUDA API](https://pytorch.org/docs/stable/cuda.html) that affords users more control over the execution of their code on CUDA devices. It would be nice if these features were available in TorchScript too!

## Motivation
Exposing streams and events at the very least would allow users to make more efficient use of their CUDA devices. At present, all CUDA operations in TorchScript are executed on the default CUDA stream, which means they are executed one at a time, and in the order in which they are submitted to the stream. Depending on the SIMD parameters with which a CUDA kernel is launched and the capabilities of the CUDA device, using only the default stream can leave hardware resources unused.

## Pitch

This issue details a proposal to expose CUDA streams and events to TorchScript users and give them the ability to specify which devices and streams their operations should execute on. Other CUDA-related APIs for memory management, communication, etc. are considered out of scope.

`c10` already has [`CUDAEvent`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/cuda/CUDAEvent.h) and [`CUDAStream`](https://github.com/pytorch/pytorch/blob/master/c10/cuda/CUDAStream.h) classes that provide a layer of indirection on top of the CUDA C/C++ API. These classes will be exposed as types in TorchScript by registering them as [custom C++ classes](https://pytorch.org/tutorials/advanced/torch_script_custom_classes.html). This will allow streams and events to be created and used in TorchScript code. They can also be returned from scripted code and used in eager mode Python code, but *users should not expect them to be fungible with streams and events created using the eager mode CUDA API*:

```
@torch.jit.script
def make_stream(device: int):
    s = torch.cuda.Stream(device)
    ...
```

To facilitate switching streams, the [`torch.cuda.stream(...)` context manager](https://github.com/pytorch/pytorch/blob/master/torch/cuda/__init__.py#L304) will be rewritten so that it is scriptable. At present, it is implemented using `@contextlib.contextmanager` which is not scriptable, so it will be rewritten as a class with `__enter__` and `__exit__`, with the generator state stored in class members, which TorchScript does support. Other unscriptable code such as `torch._C._cuda_setStream(stream._cdata)` will be replaced by functions that will call the same unscriptable code in eager mode (so as to preserve current eager mode behaviour), but will be converted by TorchScript to calls to special JIT operators that do the same thing. For example, `torch._C._cuda_setStream(stream._cdata)` can be hidden inside a function

```
# Inside torch/cuda/__init__.py
def setStream(stream):
    torch._C._cuda_setStream(stream._cdata)
```

which can then be called inside `torch.cuda.stream(...)`:
```
# Inside torch.cuda.stream
...
torch.cuda.setStream(src_prev_stream)   # used to be torch._C._cuda_setStream(stream._cdata)
...
```
which TorchScript will convert to a new custom operator `cuda::set_stream`:
```
     Operator(
         "cuda::set_stream(__torch__.torch.classes.cuda.Stream s) -> None",
         [](Stack* stack) {
           auto stream = pop(stack).toCustomClass<CUDAStream>();
           c10::cuda::setCurrentStream(stream.handle_);
         },
         aliasAnalysisFromSchema()),
```

The `device` context manager will be scripted too (as it is used by the `stream` context manager) several auxiliary operators like the one above will be added as needed.

After the completion of this work, the following code sample from the CUDA semantics documentation about working with streams should would in TorchScript too:

```
cuda = torch.device('cuda')
s = torch.cuda.Stream()  # Create a new stream.
A = torch.empty((100, 100), device=cuda).normal_(0.0, 1.0)
with torch.cuda.stream(s):
    # sum() may start execution before normal_() finishes!
    B = torch.sum(A)
```

## Alternatives

The alternative to exposing CUDA streams and events via the custom C++ class extension is to introduce them as types into the JIT type system directly. This feature is not useful on CPU, and although streams and events exist in OpenCL and HIP, they are not widely used enough to justify the additional complexity of adding new JIT types.

## Additional context

Users should anticipate small cosmetic differences between the eager mode and TorchScript CUDA APIs. For example, classes exposed via the custom C++ class extension must have all method arguments specified, even when default arguments are present on the C++ implementations of these classes. Consequently, even though a `Stream` can be created in eager mode without any arguments, both `device` and `priority` arguments will need to be provided in TorchScript.

**Comment by wconstab** [comment]
Ping @suo for design/implementation guidance

**Comment by meetshah1995** [comment]
@suo -- Is there an effort planned to support this and support of `with torch.cuda.stream(s)` context managers in TorchScript?

cc: @sidazhang

**Comment by suo** [comment]
This was previously blocked on support for `with`-statements in TorchScript, but that is available now. @SplitInfinity, do you have any thoughts on timeline here? Could be a good ramp up task for someone

**Comment by sidazhang** [comment]
@suo hey, ahh that would be pretty awesome.

I think this would get a lot of use! Do you guys have a better sense of if someone might be able to work on this?

**Comment by mkolod** [comment]
There are some details to consider here. An obvious one is whether the export to the backend runs not just on separate CUDA streams, but separate CPU threads scheduling on those streams as well.

Separate streams will only allow for using up more CUDA cores if some kernels running on a different stream have too small a grid setting to take up the whole GPU. In that case, launching on a parallel stream will take up unused SMs.

However I feel the bigger reason for multi-streaming is often that PyTorch has high kernel launch latencies. Many pointwise ops etc can take microseconds to run, but the kernel launch times are the same order of magnitude. If separate independent parts of the model all have high launch latencies but separate CPU threads enqueue on the submodels on different streams, then this may increase GPU utilization. Each kernel launch will have high launch latency anyway, but they will be interleaved due to a combination of GPU streaming and CPU multi-threading with 1 CPU thread per stream. 

See the example [here](https://gist.github.com/mkolod/fbb7c7c0ad529a19f33e6cccab44df26).

**Comment by SplitInfinity** [comment]
> Separate streams will only allow for using up more CUDA cores if some kernels running on a different stream have too small a grid setting to take up the whole GPU (little data to process, too large thread blocks to use all SMs, smaller grid/block combination due to grid-stride loops, etc.). In that case, launching on a parallel stream will take up unused SMs.

Those who use their own operators will have more control over this, but I worry that those who rely mainly on PyTorch's operator implementations might be stuck with block and grid sizes that they cannot control (well, other than forking PyTorch) and that this might make it hard to increase hardware utilization.

> An obvious one is whether the export to the backend runs not just on separate CUDA streams, but separate CPU threads scheduling on those streams as well.

Once this feature is added, I think users can exert control over that if they wish by using `torch.jit.fork` and `torch.jit.wait` to execute TorchScript code that operators on separate streams on separate threads. What do you think?

**Comment by mkolod** [comment]
@SplitInfinity These are legitimate concerns. However, short of rebuilding PyTorch to change block/grid configurations or changing custom ops, one can just observe what the launch config was using nvprof / NSight Systems / CUPTI. If it's clear that the config is suboptimal, a person tuning the code can still use streams to their advantage.

Re fork for the small kernels with big launch latencies scenario, I think you're right here. We don't really need the streams in Python for that, since they won't help due to Python's GIL situation. But fork/wait here should be perfectly fine. :)

**Comment by SplitInfinity** [comment]
I edited the issue to provide some more detail on the planned implementation

**Comment by SplitInfinity** [comment]
All, we have landed most of this in #48020. There are some small wrinkles in the API that we are trying to iron out, but it will likely take longer than expected due to the lack of a suitable code owner for the CUDA eager API that can review our changes as well as some planned but currently unfinished work on the TorchScript C++ extension API. You can have a look at the unit tests added in that PR to get a sense of the small ways in which the API differs (mostly the names of functions and arguments, there should be no semantic or behavioural differences). We'd love to get some feedback on this! Feel free to reach out via issues or Slack.

**Comment by nikithamalgifb** [comment]
Hi All, we have now landed CUDA Streams support for JIT in #53902. You can use the same CUDA eager API's to create and work with streams and events. Feel free to look at the unit tests provided in the PR to understand how the API's can be used. We'd love to get some feedback on this! Feel free to reach out via issues or Slack.
